### DISCUSSION TITLE
Wait while queries being executed

### DISCUSSION THREAD

**Discussion by NicolasFabre81** [Budibase/budibase#11277] [Help] [answered]
Hello,
I have a button that launches a lot of queries, taking up to 30 seconds or so.
I'm looking for sth to make the user wait without clicking, until it's done : spinner, timer on mouse pointer ...

Any ideas ?
@sourceae

**Comment by andz-bb** [comment, marked as answer]
hey @NicolasFabre81 

you can do this using [App State](https://docs.budibase.com/docs/app-state) and [Conditional UI](https://docs.budibase.com/docs/conditions). Here's a quick demo:

First I have set up my button actions so that before my query runs, I create a state variable `loading` and set it to `true`. Then my query runs and once it is finished I then have another step to delete the `loading` state.

<img width="947" alt="image" src="https://github.com/Budibase/budibase/assets/110921612/fd776978-d529-45f3-adec-3706916e2556">

This means that we can now conditionally render components based on that `loading` value which will only be `true` when the query is running.

The most simple example of this is that I can hide my button while it is loading by doing this:
<img width="949" alt="image" src="https://github.com/Budibase/budibase/assets/110921612/dd3d1b64-0cb1-411e-9dfa-8a1ffaed85cb">

I can then render something else in the button's place using the same conditions. I added an embed component with HTML for a Spectrum CSS spinner since all of the classes are already available in Budibase. Here's the snippet I used:

```html
  <div class="spectrum-ProgressCircle spectrum-ProgressCircle--indeterminate">
    <div class="spectrum-ProgressCircle-track"></div>
    <div class="spectrum-ProgressCircle-fills">
      <div class="spectrum-ProgressCircle-fillMask1">
        <div class="spectrum-ProgressCircle-fillSubMask1">
          <div class="spectrum-ProgressCircle-fill"></div>
        </div>
      </div>
      <div class="spectrum-ProgressCircle-fillMask2">
        <div class="spectrum-ProgressCircle-fillSubMask2">
          <div class="spectrum-ProgressCircle-fill"></div>
        </div>
      </div>
    </div>
  </div>
```

Now when I set that to conditionally show only when `loading` is `true`, the spinner replaces the button when the query is running. Here's the final result:

![loading query](https://github.com/Budibase/budibase/assets/110921612/5b105d22-2876-4646-9aab-47984cd93f15)

This is a fairly basic example but you can do a lot more using the same principles. If you don't want to hide the button another idea would be to conditionally disable the button when `loading` is `true` using the same Conditional UI drawer. Hope this helps!

**Reply by NicolasFabre81** [reply]
Great solution, thanks a lot @andz-bb 
I didn't guess it because I am still not really aware of what is possible with the embed tool.
How can one find support on such things ?
Is the spectrum CSS doc the key ? https://opensource.adobe.com/spectrum-css/index.html

**Reply by andz-bb** [reply]
hey @NicolasFabre81, the Embed component allows you to use most HTML, so it's a free area for you to add any markup as a component to your app. [Here's the docs page for the Embed component.](https://docs.budibase.com/docs/embed)

In this case for the spinner I used [this example from the Spectrum CSS docs](https://opensource.adobe.com/spectrum-css/progresscircle.html). We use Spectrum to style Budibase so those CSS classes are also available in the context of the Embed component. If you wanted to use your own CSS you could add it to the Embed component within `<style>` tags.

If you have any questions on this feel free to ask!

**Reply by NicolasFabre81** [reply]
Thanks, the key point to understand is what you say : " so those CSS classes are also available in the context of the Embed component"

**Reply by andz-bb** [reply]
@NicolasFabre81 In the embed component you can add HTML where you can style the elements using classes. Here's a really simple example:

```html
<div class="my-container">
  <p class="my-text">
    hello world
  </p>
</div>
```

by default these will appear like this:
<img width="392" alt="image" src="https://github.com/Budibase/budibase/assets/110921612/0d84dfbb-943b-4f0d-94a9-2a7c290b43d7">

but I could add styling by writing CSS for those classes like so:
```html
<style>
  .my-container {
    background: white;
  }
  .my-text {
    color: black;
    font-size: 32px;
  }
</style>
```

<img width="562" alt="image" src="https://github.com/Budibase/budibase/assets/110921612/aefec3de-24a5-4ace-a6a5-e7360ecd6717">

When I talk about using the Spectrum classes, what I mean is that in the Budibase builder and apps there is CSS already imported into the browser DOM that can be referenced inside our Embed component. This would not be the case for every CSS framework but because Budibase uses Spectrum CSS, we can reuse that for our own HTML. Since it is already being used in the core Budibase you don't need to import it or define it otherwise.

<img width="783" alt="image" src="https://github.com/Budibase/budibase/assets/110921612/bfb260b5-4a10-4339-a6f2-9092fc4af798">

<img width="335" alt="image" src="https://github.com/Budibase/budibase/assets/110921612/41ecd1d2-4ea5-4126-8821-038fd1b1a4be">
<img width="326" alt="image" src="https://github.com/Budibase/budibase/assets/110921612/8e4f1f19-8131-4fd9-bbaf-1d7c64ed65d1">

I hope this makes it a bit clearer for you! If you would like me to explain it in more detail let me know.

**Reply by NicolasFabre81** [reply]
I did understand at the first time, but your examples make it clearer indeed !
Maybe a good point would be to have a bit more doc about using css spectrum with bb.
Thanks again!
